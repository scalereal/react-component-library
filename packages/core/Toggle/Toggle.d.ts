import * as React from 'react';
import { WithStyle } from '../../utils';
import { ToggleProps } from './types';
export declare const toggleSizes: ToggleProps['size'][];
export declare const Toggle: React.FC<ToggleProps> & WithStyle;
export default Toggle;
