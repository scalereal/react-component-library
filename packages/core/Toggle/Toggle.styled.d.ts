import { ToggleProps, ToggleWrapperProps } from './types';
export declare const ToggleWrapper: import("styled-components").StyledComponent<"div", any, ToggleWrapperProps, never>;
export declare const Circle: import("styled-components").StyledComponent<"div", any, ToggleWrapperProps, never>;
export declare const Checkbox: import("styled-components").StyledComponent<"input", any, {
    type: "checkbox";
} & ToggleProps, "type">;
