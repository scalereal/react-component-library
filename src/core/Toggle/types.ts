export type ToggleSizes = 'XS' | 'S' | 'M' | 'L' | 'XL';

export interface ToggleProps {
    /** Toggle size */
    size?: ToggleSizes;
}
