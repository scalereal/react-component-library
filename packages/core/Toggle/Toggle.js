var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
import React, { useCallback } from 'react';
import * as Styled from './Toggle.styled';
export var toggleSizes = ['XS', 'S', 'M', 'L', 'XL'];
export var Toggle = React.memo(React.forwardRef(function (props, ref) {
    var size = props.size, required = props.required, onChange = props.onChange, restProps = __rest(props, ["size", "required", "onChange"]);
    var changeHandler = useCallback(function (e) {
        e.stopPropagation();
        onChange && onChange(e);
    }, [onChange]);
    return (React.createElement(Styled.ToggleWrapper, { size: size, disabled: restProps.disabled },
        React.createElement(Styled.Checkbox, __assign({ ref: ref, size: size, required: required, onChange: changeHandler }, restProps)),
        React.createElement(Styled.Circle, { size: size })));
}));
Toggle.displayName = 'Toggle';
Toggle.Style = Styled.ToggleWrapper;
Toggle.defaultProps = {
    required: false,
};
export default Toggle;
