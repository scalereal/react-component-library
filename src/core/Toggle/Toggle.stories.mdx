import { Toggle, toggleSizes } from './Toggle';
import { Preview, Story, Meta, Props, ArgsTable, Canvas } from '@storybook/addon-docs/blocks';
import { withKnobs, color, boolean, select } from '@storybook/addon-knobs';

<Meta
    title="Components/Core/Toggle"
    component={Toggle}
    argTypes={{
        size: {
            control: {
                type: 'select',
                options: Object.values(toggleSizes)
            }
        }
    }}
/>

# Toggle

Displays a Toggle

export const Template = ({ size, ...args }) => {
    return <Toggle size={size} {...args} />;
};

<Preview>
    <Story
        name="Toggle"
        args={{
            size: 'M'
        }}
    >
        {Template.bind({})}
    </Story>
</Preview>

### Sizes

There are five sizes available `XS` | `S` | `M` | `L` | `XL`.

<Preview withToolbar>
    <div style={{ display: 'flex', flexWrap: 'wrap', justifyContent: 'space-around' }}>
        <Toggle size="XS" />
        <Toggle size="S" />
        <Toggle size="M" />
        <Toggle size="L" />
        <Toggle size="XL" />
    </div>
</Preview>

### Disabled

<Preview withToolbar>
    <div>
        <Toggle size="M" disabled={true} />
    </div>
</Preview>

### Props

<Props story="Toggle" />
