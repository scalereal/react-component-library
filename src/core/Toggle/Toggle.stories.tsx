import { Meta, Story } from '@storybook/react/types-6-0';
import React from 'react';
import Toggle from "./Toggle"

export default {
    title: 'Components/Core/Toggle',
    component: Toggle,
} as Meta

const Template:Story = ({ size, ...args}) => {
    return <Toggle size={size} {...args} />
}

export const Default = Template.bind({});
Default.args = {
    size:'M'
}
