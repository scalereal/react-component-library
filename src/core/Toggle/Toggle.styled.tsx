import styled from "styled-components"
import { toggleSizes } from "../../theme/sizes"
import { ToggleProps } from './types';
import { defaultTheme } from "../../theme/theme"

const getHeight = ({ size }: ToggleProps) => (size ? toggleSizes[size] : toggleSizes['M']);
const getWidth = (props: ToggleProps) => `calc(${getHeight(props)} * 2)`;

export const Wrapper = styled('div')<ToggleProps>`
    position: relative;
    width: ${getWidth};
    height: ${getHeight};
    border-radius: 15px;

    &:focus-within {
        box-shadow: 0 0 0 1pt ${defaultTheme.whiteColor};
    }

    & > * {
        width: 100%;
        height: 100%;
        border-radius: 15px;
    }
`;
export const Circle = styled('div')<ToggleProps>`
    z-index: 1;
    position: absolute;
    background-color: ${defaultTheme.shadowColor};

    &::after {
        content: '';
        margin: 3px;
        display: block;
        border-radius: 50%;
        background: ${defaultTheme.whiteColor};
        transition: 0.2s;
        width: ${props => `calc(${getHeight(props)} - 6px) `};
        height: ${props => `calc(${getHeight(props)} - 6px) `};
    }
`;

export const Checkbox = styled('input').attrs({ type: 'checkbox' })<ToggleProps>`
    opacity: 0;
    z-index: 2;
    margin: 0;
    position: absolute;
    cursor: pointer;

    &:focus {
        outline: none;
    }

    &:disabled {
        cursor: not-allowed;
        & + ${Circle} {
            background: ${defaultTheme.shadowColor};
        }
    }

    &:checked + ${Circle} {
        background: ${defaultTheme.blackColor};
        &::after {
            margin-left: ${props => `calc(${getWidth(props)} - (${getHeight(props)}) + 3px) `};
        }
    }
`;
