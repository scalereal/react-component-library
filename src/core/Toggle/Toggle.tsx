import * as React from 'react'
import { WithStyle } from '../../utils'
import * as Styled from './Toggle.styled'
import { ToggleProps } from './types'

export const toggleSizes : ToggleProps['size'][] = ['XS', 'S', 'M', 'L', 'XL']

export const Toggle: React.FC<ToggleProps> & WithStyle = React.memo(
    React.forwardRef((props, ref) => {
        const { size, required, onChange, ...restProps } = props;

        const changeHandler = React.useCallback(
            (e: any) => {
                e.stopPropagation();
                onChange && onChange(e);
            },
            [onChange]
        );

        return (
                <Styled.ToggleWrapper size={size} disabled={restProps.disabled}>
                    {
                        <Styled.Checkbox ref={ref}  size={size} required={required} onChange={changeHandler} {...restProps} />
                    }
                    <Styled.Circle size={size} />
                </Styled.ToggleWrapper>
        );
    })
);

Toggle.displayName = 'Toggle';
Toggle.Style = Styled.ToggleWrapper;
Toggle.defaultProps = {
    required: false,
};

export default Toggle
