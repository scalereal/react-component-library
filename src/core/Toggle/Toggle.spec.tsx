import { render } from '@testing-library/react';
import React from 'react';
import Toggle from './Toggle';

describe('Toggle', () => {
    it('should render correctly with all the props given', () => {
        const { container } = render(<Toggle size="M" />);
        expect(container).toMatchSnapshot();
    });
});
