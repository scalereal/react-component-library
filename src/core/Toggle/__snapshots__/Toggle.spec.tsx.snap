// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle should render correctly with all the props given 1`] = `
<div>
  <div
    class="sc-bdfBwQ ljahyS"
  >
    <input
      class="sc-dlfnbm hWcVOc"
      type="checkbox"
    />
    <div
      class="sc-gsTCUz jWJZQO"
    />
  </div>
</div>
`;
